### ISSUE TITLE
Assertion failed in destructor of `FileSegmentsHolder`

### ISSUE THREAD

**Issue by CurtizJ** [ClickHouse/ClickHouse#36789]
**Describe the bug**
https://s3.amazonaws.com/clickhouse-test-reports/36077/20e2fa9b3f12e03f888461369bb0f78bc5709f60/stress_test__debug__actions_.html

```
2022.04.29 00:12:50.162735 [ 4058 ] {} <Fatal> BaseDaemon: ########################################
2022.04.29 00:12:50.250985 [ 4058 ] {} <Fatal> BaseDaemon: (version 22.5.1.1, build id: 8E6CC4881A822B3B) (from thread 760) (query_id: eb97d5a7-b931-45cb-a882-86dafb9840cc::201403_17_24_1) (query: ) Received signal Aborted (6)
2022.04.29 00:12:50.578744 [ 4058 ] {} <Fatal> BaseDaemon:
2022.04.29 00:12:50.582239 [ 4058 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f5de2a9803b 0x7f5de2a77859 0x7f5de2a77729 0x7f5de2a89006 0x258d71a1 0x258d744a 0x262966a7 0x26296675 0x26296655 0x26296635 0x26296615 0x262965f5 0x262965d5 0x26295fc1 0x26276847 0x262767d5 0x262767b9 0x262765ae 0x17b65331 0x17b652d9 0x258c8b4c 0x26258031 0x262582fa 0x2625824b 0x262587ca 0x17b65331 0x17b652d9 0x262aa5ac 0x262abba5 0x262abbd9 0x2620e58c 0x2620e4fc 0x2620e499 0x2620e035 0x262abe35 0x262abe59 0x2620e58c 0x2620e4fc 0x2620e499 0x2620e035 0x2620e0ae
2022.04.29 00:12:50.686906 [ 4058 ] {} <Fatal> BaseDaemon: 4. raise @ 0x7f5de2a9803b in ?
2022.04.29 00:12:50.727429 [ 4058 ] {} <Fatal> BaseDaemon: 5. abort @ 0x7f5de2a77859 in ?
2022.04.29 00:12:50.743583 [ 4058 ] {} <Fatal> BaseDaemon: 6. ? @ 0x7f5de2a77729 in ?
2022.04.29 00:12:50.746819 [ 4058 ] {} <Fatal> BaseDaemon: 7. ? @ 0x7f5de2a89006 in ?
2022.04.29 00:12:51.217785 [ 4058 ] {} <Fatal> BaseDaemon: 8. /build/build_docker/../src/Common/FileSegment.cpp:655: DB::FileSegment::assertDetachedStatus() const @ 0x258d71a1 in /usr/bin/clickhouse
2022.04.29 00:12:51.624284 [ 4058 ] {} <Fatal> BaseDaemon: 9. /build/build_docker/../src/Common/FileSegment.cpp:694: DB::FileSegmentsHolder::~FileSegmentsHolder() @ 0x258d744a in /usr/bin/clickhouse
2022.04.29 00:12:52.266279 [ 4058 ] {} <Fatal> BaseDaemon: 10. /build/build_docker/../contrib/libcxx/include/optional:239: std::__1::__optional_destruct_base<DB::FileSegmentsHolder, false>::~__optional_destruct_base() @ 0x262966a7 in /usr/bin/clickhouse
2022.04.29 00:12:52.769433 [ 4058 ] {} <Fatal> BaseDaemon: 11. /build/build_docker/../contrib/libcxx/include/optional:312: std::__1::__optional_storage_base<DB::FileSegmentsHolder, false>::~__optional_storage_base() @ 0x26296675 in /usr/bin/clickhouse
2022.04.29 00:12:53.329024 [ 4058 ] {} <Fatal> BaseDaemon: 12. /build/build_docker/../contrib/libcxx/include/optional:485: std::__1::__optional_copy_base<DB::FileSegmentsHolder, false>::~__optional_copy_base() @ 0x26296655 in /usr/bin/clickhouse
2022.04.29 00:12:53.875401 [ 4058 ] {} <Fatal> BaseDaemon: 13. /build/build_docker/../contrib/libcxx/include/optional:513: std::__1::__optional_move_base<DB::FileSegmentsHolder, false>::~__optional_move_base() @ 0x26296635 in /usr/bin/clickhouse
2022.04.29 00:12:54.420383 [ 4058 ] {} <Fatal> BaseDaemon: 14. /build/build_docker/../contrib/libcxx/include/optional:546: std::__1::__optional_copy_assign_base<DB::FileSegmentsHolder, false>::~__optional_copy_assign_base() @ 0x26296615 in /usr/bin/clickhouse
2022.04.29 00:12:54.785569 [ 4058 ] {} <Fatal> BaseDaemon: 15. /build/build_docker/../contrib/libcxx/include/optional:578: std::__1::__optional_move_assign_base<DB::FileSegmentsHolder, false>::~__optional_move_assign_base() @ 0x262965f5 in /usr/bin/clickhouse
2022.04.29 00:12:55.418670 [ 4058 ] {} <Fatal> BaseDaemon: 16. /build/build_docker/../contrib/libcxx/include/optional:621: std::__1::optional<DB::FileSegmentsHolder>::~optional() @ 0x262965d5 in /usr/bin/clickhouse
2022.04.29 00:12:55.942669 [ 4058 ] {} <Fatal> BaseDaemon: 17. /build/build_docker/../src/Disks/IO/CachedReadBufferFromRemoteFS.h:12: DB::CachedReadBufferFromRemoteFS::~CachedReadBufferFromRemoteFS() @ 0x26295fc1 in /usr/bin/clickhouse
2022.04.29 00:12:56.188207 [ 4058 ] {} <Fatal> BaseDaemon: 18. /build/build_docker/../contrib/libcxx/include/__memory/construct_at.h:57: void std::__1::__destroy_at<DB::CachedReadBufferFromRemoteFS, 0>(DB::CachedReadBufferFromRemoteFS*) @ 0x26276847 in /usr/bin/clickhouse
2022.04.29 00:12:56.932843 [ 4058 ] {} <Fatal> BaseDaemon: 19. /build/build_docker/../contrib/libcxx/include/__memory/construct_at.h:82: void std::__1::destroy_at<DB::CachedReadBufferFromRemoteFS, 0>(DB::CachedReadBufferFromRemoteFS*) @ 0x262767d5 in /usr/bin/clickhouse
2022.04.29 00:12:57.214613 [ 4058 ] {} <Fatal> BaseDaemon: 20. /build/build_docker/../contrib/libcxx/include/__memory/allocator_traits.h:321: void std::__1::allocator_traits<std::__1::allocator<DB::CachedReadBufferFromRemoteFS> >::destroy<DB::CachedReadBufferFromRemoteFS, void, void>(std::__1::allocator<DB::CachedReadBufferFromRemoteFS>&, DB::CachedReadBufferFromRemoteFS*) @ 0x262767b9 in /usr/bin/clickhouse
2022.04.29 00:12:57.592011 [ 4058 ] {} <Fatal> BaseDaemon: 21. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:310: std::__1::__shared_ptr_emplace<DB::CachedReadBufferFromRemoteFS, std::__1::allocator<DB::CachedReadBufferFromRemoteFS> >::__on_zero_shared() @ 0x262765ae in /usr/bin/clickhouse
2022.04.29 00:12:57.703989 [ 4058 ] {} <Fatal> BaseDaemon: 22. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:175: std::__1::__shared_count::__release_shared() @ 0x17b65331 in /usr/bin/clickhouse
2022.04.29 00:12:57.839772 [ 4058 ] {} <Fatal> BaseDaemon: 23. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:216: std::__1::__shared_weak_count::__release_shared() @ 0x17b652d9 in /usr/bin/clickhouse
2022.04.29 00:12:58.396533 [ 4058 ] {} <Fatal> BaseDaemon: 24. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:704: std::__1::shared_ptr<DB::SeekableReadBuffer>::~shared_ptr() @ 0x258c8b4c in /usr/bin/clickhouse
2022.04.29 00:12:59.035949 [ 4058 ] {} <Fatal> BaseDaemon: 25. /build/build_docker/../src/Disks/IO/ReadBufferFromRemoteFSGather.h:23: DB::ReadBufferFromRemoteFSGather::~ReadBufferFromRemoteFSGather() @ 0x26258031 in /usr/bin/clickhouse
2022.04.29 00:12:59.898668 [ 4058 ] {} <Fatal> BaseDaemon: 26. /build/build_docker/../src/Disks/IO/ReadBufferFromRemoteFSGather.h:100: DB::ReadBufferFromS3Gather::~ReadBufferFromS3Gather() @ 0x262582fa in /usr/bin/clickhouse
2022.04.29 00:13:01.652061 [ 4058 ] {} <Fatal> BaseDaemon: 27. /build/build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:54: std::__1::default_delete<DB::ReadBufferFromS3Gather>::operator()(DB::ReadBufferFromS3Gather*) const @ 0x2625824b in /usr/bin/clickhouse
2022.04.29 00:13:02.164544 [ 4058 ] {} <Fatal> BaseDaemon: 28. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:264: std::__1::__shared_ptr_pointer<DB::ReadBufferFromS3Gather*, std::__1::default_delete<DB::ReadBufferFromS3Gather>, std::__1::allocator<DB::ReadBufferFromS3Gather> >::__on_zero_shared() @ 0x262587ca in /usr/bin/clickhouse
2022.04.29 00:13:02.426634 [ 4058 ] {} <Fatal> BaseDaemon: 29. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:175: std::__1::__shared_count::__release_shared() @ 0x17b65331 in /usr/bin/clickhouse
2022.04.29 00:13:02.614435 [ 4058 ] {} <Fatal> BaseDaemon: 30. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:216: std::__1::__shared_weak_count::__release_shared() @ 0x17b652d9 in /usr/bin/clickhouse
2022.04.29 00:13:02.867516 [ 4058 ] {} <Fatal> BaseDaemon: 31. /build/build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:704: std::__1::shared_ptr<DB::ReadBufferFromRemoteFSGather>::~shared_ptr() @ 0x262aa5ac in /usr/bin/clickhouse
2022.04.29 00:13:03.201874 [ 4058 ] {} <Fatal> BaseDaemon: 32. /build/build_docker/../src/Disks/IO/ReadIndirectBufferFromRemoteFS.h:18: DB::ReadIndirectBufferFromRemoteFS::~ReadIndirectBufferFromRemoteFS() @ 0x262abba5 in /usr/bin/clickhouse
2022.04.29 00:13:03.398391 [ 4058 ] {} <Fatal> BaseDaemon: 33. /build/build_docker/../src/Disks/IO/ReadIndirectBufferFromRemoteFS.h:18: DB::ReadIndirectBufferFromRemoteFS::~ReadIndirectBufferFromRemoteFS() @ 0x262abbd9 in /usr/bin/clickhouse
2022.04.29 00:13:03.711699 [ 4058 ] {} <Fatal> BaseDaemon: 34. /build/build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:55: std::__1::default_delete<DB::SeekableReadBuffer>::operator()(DB::SeekableReadBuffer*) const @ 0x2620e58c in /usr/bin/clickhouse
2022.04.29 00:13:04.114308 [ 653 ] {} <Fatal> Application: Child process was terminated by signal 6.
```

**How to reproduce**
The assertion failed during merge. Related logs are attached.
[log_assert.txt](https://github.com/ClickHouse/ClickHouse/files/8590290/log_assert.txt)

**Comment by tavplubix** [comment]
AFAIK it's fixed in #36660 and #36743

**Comment by Algunenano** [comment]
I don't think this is fixed. It has appeared in https://github.com/ClickHouse/ClickHouse/pull/36884 CI which uses master from c21cff79059a617f70b2f4e421a3eb3132afa58e


Stress test run: https://s3.amazonaws.com/clickhouse-test-reports/36884/5190d9e85a8633ae415b077154f24836b44e92b0/stress_test__debug__actions_.html
